File: bc2025/src/philip_06/utils/Actions.cpp
#include "utils/Actions.hpp"
#include "utils/Constants.hpp"
#include "RobotPlayer.hpp"
#include "Helper/Comms.hpp"
#include "roles/Mopper.hpp"
#include "roles/Tower.hpp"

namespace philip_06 {
namespace utils {

std::unordered_map<MapLocation, int> Actions::badRuins;

bool Actions::spawn(RobotController rc, UnitType type, MapLocation dest) {
    // RobotPlayer::rng is api.hpp's ::Random, which has no nextBoolean();
    // java.util.Random.nextBoolean() is `next(1) != 0`, i.e. a fair coin, so
    // `nextInt(2) != 0` is the faithful stand-in (the RNG stream already
    // differs from java.util.Random's - see CONVENTIONS.md).
    if (dest == MapLocation(61, 61) && !roles::Tower::target.isNull() &&
        RobotPlayer::rng.nextInt(2) != 0)
        dest = roles::Tower::target;
    for (const Direction& dir : Direction::allDirections()) {
        if (rc.canBuildRobot(type, rc.getLocation().add(dir))) {
            rc.buildRobot(type, rc.getLocation().add(dir));
            if (rc.canSendMessage(rc.getLocation().add(dir))) {
                rc.sendMessage(rc.getLocation().add(dir), Helper::Comms::encodeLocation(dest));
            }
            return true;
        }
    }
    return false;
}

void Actions::makeTower(RobotController rc, MapLocation ruin) {
    if (ruin.isNull()) return;
    if (rc.getNumberTowers() == GameConstants::MAX_NUMBER_OF_TOWERS) return;
    if (badRuins.count(ruin)) {
        if (rc.getRoundNum() - badRuins.at(ruin) < 10)
            return;
        else
            badRuins.erase(ruin);
    }
    if (rc.getLocation().distanceSquaredTo(ruin) > 9) {
        roles::Mopper::moveTowardsMindlessly(rc, ruin);
        return;
    }
    // Java `UnitType type = null;` - UnitType is an enum class here, so the
    // null state is tracked by a companion flag; `type` is only read once
    // typeIsNull is false.
    UnitType type = UnitType::SOLDIER;
    bool typeIsNull = true;
    if (isAlly(rc.senseMapInfo(ruin.add(Direction::all[Direction::NORTH])).getMark())) {
        type = UnitType::LEVEL_ONE_PAINT_TOWER;
        typeIsNull = false;
    } else if (isAlly(rc.senseMapInfo(ruin.add(Direction::all[Direction::EAST])).getMark())) {
        type = UnitType::LEVEL_ONE_MONEY_TOWER;
        typeIsNull = false;
    } else if (isAlly(rc.senseMapInfo(ruin.add(Direction::all[Direction::SOUTH])).getMark())) {
        type = UnitType::LEVEL_ONE_DEFENSE_TOWER;
        typeIsNull = false;
    }
    if (typeIsNull) {
        if (RobotPlayer::rng.nextInt(3) == 1 || rc.getMoney() > 5000) {
            type = UnitType::LEVEL_ONE_PAINT_TOWER;
            if (rc.canMark(ruin.add(Direction::all[Direction::NORTH])))
                rc.mark(ruin.add(Direction::all[Direction::NORTH]), false);
        } else {
            type = UnitType::LEVEL_ONE_MONEY_TOWER;
            if (rc.canMark(ruin.add(Direction::all[Direction::EAST])))
                rc.mark(ruin.add(Direction::all[Direction::EAST]), false);
        }
    }
    MapLocation best = MapLocation::NONE;
    int _cost = 0;
    bool _colour = false;
    vector<vector<bool>> arr = rc.getTowerPattern(type);
    for (int i = -2; i <= 2; i++) {
        for (int j = -2; j <= 2; j++) {
            if (i == 0 && j == 0) continue;
            if (rc.canSenseLocation(ruin.translate(i, j))) {
                PaintType cur = rc.senseMapInfo(ruin.translate(i, j)).getPaint();
                if (cur == PaintType::EMPTY ||
                    (cur == PaintType::ALLY_PRIMARY && arr[i + 2][j + 2]) ||
                    (cur == PaintType::ALLY_SECONDARY && !arr[i + 2][j + 2])) {
                    _cost++;
                    if (best.isNull() || rc.getLocation().distanceSquaredTo(best) >
                                             rc.getLocation().distanceSquaredTo(ruin.translate(i, j))) {
                        best = ruin.translate(i, j);
                        _colour = arr[i + 2][j + 2];
                    }
                } else if (cur == PaintType::ENEMY_PRIMARY || cur == PaintType::ENEMY_SECONDARY) {
                    _cost = 1000000;
                }
            }
        }
    }
    if (_cost >= 25) {
        badRuins[ruin] = rc.getRoundNum();
    } else if (!best.isNull()) {
        rc.setIndicatorDot(best, 255, 0, 0);
        if (rc.canAttack(best))
            rc.attack(best, _colour);
    }
    if (rc.senseNearbyRobots(ruin, rc.getLocation().distanceSquaredTo(ruin), rc.getTeam())
            .empty()) {
        if (!best.isNull() && rc.getLocation().distanceSquaredTo(best) > 9)
            roles::Mopper::moveTowardsMindlessly(rc, best);
        else
            roles::Mopper::moveTowardsMindlessly(rc, ruin);
    }
    if (rc.canCompleteTowerPattern(type, ruin))
        rc.completeTowerPattern(type, ruin);
}

void Actions::getPaint(RobotController rc) {
    if (rc.getHealth() <= 40) return;
    for (const MapLocation& loc : rc.senseNearbyRuins(-1)) {
        RobotInfo bot = rc.senseRobotAtLocation(loc);
        // RobotInfo has no NONE sentinel in api.hpp; a default RobotInfo's
        // location is MapLocation::NONE, used here as the null proxy.
        if (bot.getLocation().isNull()) continue;
        if (rc.canTransferPaint(
                loc, -std::min(Constants::paintCapacity(rc.getType()) - rc.getPaint(),
                               bot.paintAmount)))
            rc.transferPaint(loc, -std::min(Constants::paintCapacity(rc.getType()) - rc.getPaint(),
                                            bot.paintAmount));
    }
}

}  // namespace utils
}  // namespace philip_06

File: py/bc2025/src/philip_06/utils/Actions.py
# Port of philip_06/utils/Actions.java. See py/CONVENTIONS.md.
# Bot-internal imports use the module form (`import X` / `from pkg import X`)
# and sit below the class because this bot's import graph is cyclic - names
# resolve at call time, like the reechee port's `import RobotPlayer` +
# `RobotPlayer.RobotPlayer.*`.
from api import *

# Java: import java.util.HashMap -> dict


class Actions:
    @staticmethod
    def spawn(rc, type, dest):
        # Java: RobotPlayer.rng.nextBoolean() - api.Random (the java.util.Random
        # stand-in) has no nextBoolean(); Java's is a fair coin (next(1) != 0),
        # so nextInt(2) != 0 is the faithful stand-in (same as the C++ port).
        if dest == MapLocation(61, 61) and Tower.Tower.target is not None and RobotPlayer.RobotPlayer.rng.nextInt(2) != 0:
            dest = Tower.Tower.target
        for dir in Direction.allDirections():
            if rc.canBuildRobot(type, rc.getLocation().add(dir)):
                rc.buildRobot(type, rc.getLocation().add(dir))
                if rc.canSendMessage(rc.getLocation().add(dir)):
                    rc.sendMessage(rc.getLocation().add(dir), Comms.Comms.encodeLocation(dest))
                return True
        return False

    badRuins = {}  # static HashMap<MapLocation,Integer>

    @staticmethod
    def makeTower(rc, ruin):
        if ruin is None:
            return
        if rc.getNumberTowers() == GameConstants.MAX_NUMBER_OF_TOWERS:
            return
        if ruin in Actions.badRuins:
            if rc.getRoundNum() - Actions.badRuins[ruin] < 10:
                return
            else:
                del Actions.badRuins[ruin]
        if rc.getLocation().distanceSquaredTo(ruin) > 9:
            Mopper.Mopper.moveTowardsMindlessly(rc, ruin)
            return
        type = None
        if rc.senseMapInfo(ruin.add(Direction.NORTH)).getMark().isAlly():
            type = UnitType.LEVEL_ONE_PAINT_TOWER
        elif rc.senseMapInfo(ruin.add(Direction.EAST)).getMark().isAlly():
            type = UnitType.LEVEL_ONE_MONEY_TOWER
        elif rc.senseMapInfo(ruin.add(Direction.SOUTH)).getMark().isAlly():
            type = UnitType.LEVEL_ONE_DEFENSE_TOWER
        if type is None:
            if RobotPlayer.RobotPlayer.rng.nextInt(3) == 1 or rc.getMoney() > 5000:
                type = UnitType.LEVEL_ONE_PAINT_TOWER
                if rc.canMark(ruin.add(Direction.NORTH)):
                    rc.mark(ruin.add(Direction.NORTH), False)
            else:
                type = UnitType.LEVEL_ONE_MONEY_TOWER
                if rc.canMark(ruin.add(Direction.EAST)):
                    rc.mark(ruin.add(Direction.EAST), False)
        best = None
        _cost = 0
        _colour = False
        arr = rc.getTowerPattern(type)
        for i in range(-2, 3):
            for j in range(-2, 3):
                if i == 0 and j == 0:
                    continue
                if rc.canSenseLocation(ruin.translate(i, j)):
                    cur = rc.senseMapInfo(ruin.translate(i, j)).getPaint()
                    if cur == PaintType.EMPTY or (cur == PaintType.ALLY_PRIMARY and arr[i + 2][j + 2]) or (cur == PaintType.ALLY_SECONDARY and not arr[i + 2][j + 2]):
                        _cost += 1
                        if best is None or rc.getLocation().distanceSquaredTo(best) > rc.getLocation().distanceSquaredTo(ruin.translate(i, j)):
                            best = ruin.translate(i, j)
                            _colour = arr[i + 2][j + 2]
                    elif cur == PaintType.ENEMY_PRIMARY or cur == PaintType.ENEMY_SECONDARY:
                        _cost = 1000000
        if _cost >= 25:
            Actions.badRuins[ruin] = rc.getRoundNum()
        elif best is not None:
            rc.setIndicatorDot(best, 255, 0, 0)
            if rc.canAttack(best):
                rc.attack(best, _colour)
        if len(rc.senseNearbyRobots(ruin, rc.getLocation().distanceSquaredTo(ruin), rc.getTeam())) == 0:
            if best is not None and rc.getLocation().distanceSquaredTo(best) > 9:
                Mopper.Mopper.moveTowardsMindlessly(rc, best)
            else:
                Mopper.Mopper.moveTowardsMindlessly(rc, ruin)
        if rc.canCompleteTowerPattern(type, ruin):
            rc.completeTowerPattern(type, ruin)

    @staticmethod
    def getPaint(rc):
        if rc.getHealth() <= 40:
            return
        for loc in rc.senseNearbyRuins(-1):
            bot = rc.senseRobotAtLocation(loc)
            if bot is None:
                continue
            if rc.canTransferPaint(loc, -min(rc.getType().paintCapacity - rc.getPaint(), bot.paintAmount)):
                rc.transferPaint(loc, -min(rc.getType().paintCapacity - rc.getPaint(), bot.paintAmount))


# Java's file-level imports, deferred below the class: several sibling modules
# do `from utils.Actions import Actions`, which only resolves once the class
# object already exists in the partially-initialised module.
import RobotPlayer
from Helper import Comms
from Helper import Pathfind  # imported by the Java file, unused in this file
from roles import Mopper
from roles import Tower
